stage1/init/common/pod.go

```diff
@@ -565,6 +565,56 @@ func PodToSystemd(p *stage1commontypes.Pod, interactive bool, flavor string, pri
 	return nil
 }
 
+// evaluateAppMountPath tries to resolve symlinks within the path.
+// It returns the actual relative path for the given path.
+// TODO(yifan): This is a temporary fix for systemd-nspawn not handling symlink mounts well.
+// Could be removed when https://github.com/systemd/systemd/issues/2860 is resolved, and systemd
+// version is bumped.
+func evaluateAppMountPath(appRootfs, path string) (string, error) {
+	link := appRootfs
+
+	paths := strings.Split(path, "/")
+	for i, p := range paths {
+		next := filepath.Join(link, p)
+
+		if !strings.HasPrefix(next, appRootfs) {
+			return "", fmt.Errorf("path escapes app's root: %q", path)
+		}
+
+		fi, err := os.Lstat(next)
+		if err != nil {
+			if os.IsNotExist(err) {
+				link = filepath.Join(append([]string{link}, paths[i:]...)...)
+				break
+			}
+			return "", err
+		}
+
+		if fi.Mode()&os.ModeType != os.ModeSymlink {
+			link = filepath.Join(link, p)
+			continue
+		}
+
+		// Evaluate the symlink.
+		target, err := os.Readlink(next)
+		if err != nil {
+			return "", err
+		}
+
+		if filepath.IsAbs(target) {
+			link = filepath.Join(appRootfs, target)
+		} else {
+			link = filepath.Join(link, target)
+		}
+
+		if !strings.HasPrefix(link, appRootfs) {
+			return "", fmt.Errorf("symlink %q escapes app's root with value %q", next, target)
```

Comment:
Ditto, this would avoid an error if the image has silly symlinks but it's probably not needed right now.